### PR TITLE
Large CSV details

### LINKED ISSUE #3961: Grid Viewer Prerformance Description
@willingc @jasongrout Based on our old talks and the discussion at PyCaribbean we should explicitly describe in the docs and in the demo the performance capabilities of the grid viewer 

See https://github.com/jupyterlab/jupyterlab-demo/issues/63

### PR DESCRIPTION
**Pull request by jzf2101**
@jasongrout closes #3961 

#3997

### CODE REVIEW

File: docs/source/user/file_formats.rst [hunk 1]
```diff
@@ -127,6 +127,10 @@ browser and select the “Editor” item in the “Open With” submenu:
     <iframe src="https://www.youtube-nocookie.com/embed/b5oAoVB3Wd4?rel=0&amp;showinfo=0" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>
   </div>
 
+Our powerful grid viewer can handle large CSVs.  The `demo <https://mybinder.org/v2/gh/jupyterlab/jupyterlab-demo/18a9793b58ba86660b5ab964e1aeaf7324d667c8?urlpath=lab%2Ftree%2Fdemo%2FLorenz.ipynb>`__ contains a 200MB public transportation file from Geneva named ``big.csv`` with ~1.2 million rows.
+To see how JupyterLab can handle even larger datasets, `click on this example on the PhospohorJS website <http://phosphorjs.github.io/examples/datagrid/index.html>`__ to see a
+trillion row, trillion column CSV example.
```
Comment (sccolbert) [inline docs/source/user/file_formats.rst]: That grid is fake generated data. It's not actually CSV. I would just strike "CSV" from this sentence.

**Comment by jasongrout** [comment]
The trillion row example isn't in JLab. I think it makes sense to add a note about it to the demo, as an example of what the underlying phosphor grid can do, but not advertise it here until we have large examples like that working in JLab.

**Comment by jzf2101** [comment]
Let me rephrase it then

**Comment by jasongrout** [comment]
None of the other examples points out a specific example file (before, we were going for clarity and brevity, I thought). In my opinion, it feels a bit verbose to have this example file called out, especially when we just said it handled 250mb csv files in the previous paragraph. Perhaps 

Also, as this is user-focused documentation about JupyterLab - I think introducing the name "PhosphorJS" and pointing out the trillion row example will be confusing to users.  I don't think most users reading this introductory documentation will understand the distinction between what Phosphor supports and what JupyterLab supports, and it will set up the expectation for them that JupyterLab supports trillions of rows when it doesn't (yet), despite how carefully we phrase it.

I'd prefer we wait until we actually support huge files in JupyterLab itself using streaming (so hundreds of millions of rows? billions of rows?) before mentioning capabilities like this in the JupyterLab introductory user documentation. *Then* let's go to town advertising this capability in JupyterLab!

**Comment by jasongrout** [comment]
Perhaps another way we can mention phosphor while still keeping it *very* clear that we are not talking about jupyterlab is to say something like: `The default viewer for CSV files in JupyterLab is a
high-performance data grid viewer, based on the [PhosphorJS grid viewer](link), which can also handle tab- and
semicolon-separated values.` where the link is *not* to the demo, but to an appropriate documentation page for the phosphor grid. (I think linking to the demo will introduce the same confusion I talked about above, while linking to a phosphor documentation page about the grid more clearly separates it as a separate project.) @sccolbert - do you have a documentation page for the grid?

**Comment by jzf2101** [comment]
@jasongrout I disagree since we frequently get asked questions about how much data the CSV viewer can handle and also showing off the CSV viewer's capabilities with large files has been a key point of showing off JLab

**Comment by jasongrout** [comment]
> I disagree since we frequently get asked questions about how much data the CSV viewer can handle and also showing off the CSV viewer's capabilities with large files has been a key point of showing off JLab

There are two points in my comments above: (a) the link to the demo file is probably unnecessary since we are already telling them that jlab handles files that big, and we don't have example files for any of the other formats, and one of our original goals in this user walkthrough documentation was to be fairly concise anyway, and the second point was (b) we should be *very* careful and clear about what JupyterLab itself supports, and the distinction between what the phosphor grid supports and what jupyterlab supports will likely be lost on most people, leading to people expecting jupyterlab to support trillions of rows, which it does not support currently.

I definitely agree that we should include some text about how large of files the csv viewer can currently handle. Thanks to your suggestion, I changed the docs in #3997 to say: "JupyterLab can open files up to the maximum string size in the browser (which ranges from approximately 250MB to 2GB, depending on the browser)". I thought this addressed the questions about how much data the CSV viewer could handle - are you disagreeing that this is clear enough? If you feel we really should include a demo file, what do you think about making it more concise with a "like [this file](link to file)" remark in the sentence above.

Or are you disagreeing with point (b)?

**Comment by jzf2101** [comment]
I'm going to update this with respect to the notes you've made @jasongrout

**Comment by jzf2101** [comment]
@jasongrout it's ready to review again

File: docs/source/user/file_formats.rst [hunk 2]
```diff
@@ -101,18 +101,16 @@ submenu:
...
 
-Files with rows of comma-separated values (CSV files) are a common format for
-tabular data. The default viewer for CSV files in JupyterLab is a
-high-performance data grid viewer (which can also handle tab- and
-semicolon-separated values). JupyterLab can open files up to the maximum string
-size in the browser (which ranges from approximately 250MB to 2GB, depending on
-the browser):
+Files with rows of delimiter-separated values (CSV files, TSV files) are a common format for
+tabular data. The default viewer for these files in JupyterLab is a
+high-performance data grid viewer which can handles comma-separated, tab-separated, and
```
Comment (jasongrout) [inline docs/source/user/file_formats.rst]: can handle (delete s)

File: docs/source/user/file_formats.rst [hunk 3]
```diff
@@ -129,6 +127,24 @@ browser and select the “Editor” item in the “Open With” submenu:
     <iframe src="https://www.youtube-nocookie.com/embed/b5oAoVB3Wd4?rel=0&amp;showinfo=0" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>
   </div>
 
+JupyterLab's powerful grid viewer can load files up to the maximum string
```
Comment (jasongrout) [inline docs/source/user/file_formats.rst]: I would delete "powerful" - I think it's sounds too self-aggrandizing in the docs, and would rather let the *user* say "wow, that's powerful"

File: docs/source/user/file_formats.rst [hunk 4]
```diff
@@ -129,6 +127,24 @@ browser and select the “Editor” item in the “Open With” submenu:
...
+size in the browser. Our experience testing the grid viewer on various browsers suggests that maximum string size varies with browser.
+Below is a table that reflects the maximum size we've successfully tested the grid viewer by browser:
```
Comment (jasongrout) [inline docs/source/user/file_formats.rst]: I would delete the "Our experience testing the grid viewer ..." sentence. We said it was a browser issue in the previous sentence, and the next sentence and table very clearly indicate it depends on the browser.

Comment (jasongrout) [inline docs/source/user/file_formats.rst]: "JupyterLab's grid viewer can open large files, up to the maximum string size of the particular browser. Below is a table that shows the sizes of the largest test files we successfully opened in each browser."

* I would delete "powerful" - I think that term is relative, and feels a *bit* self-aggrandizing, and would rather just stick to the facts and let the *user* say "wow, that's powerful!".
* I combined the second sentence into the other two.

File: docs/source/user/file_formats.rst [hunk 5]
```diff
@@ -129,6 +127,24 @@ browser and select the “Editor” item in the “Open With” submenu:
...
++---------+----------+
+| Browser | Max Size |
++=========+==========+
+| Firefox | ~200MB   |
++---------+----------+
+| Chrome  | ~700MB   |
++---------+----------+
+| Safari  | ~1.5GB   |
++---------+----------+
+| IE      | ~1.5GB   |
```
Comment (jasongrout) [inline docs/source/user/file_formats.rst]: We don't support IE in JLab for other reasons, so let's remove this row.

File: docs/source/user/file_formats.rst [hunk 6]
```diff
@@ -129,6 +127,24 @@ browser and select the “Editor” item in the “Open With” submenu:
...
++---------+----------+
+
+These numbers are intented to be conservative estimates.  Additionally, string encoding will also affect the ability to load CSV-like files on JupyterLab.
```
Comment (jasongrout) [inline docs/source/user/file_formats.rst]: "These numbers are conservative estimates. The actual sizes of files that can be successfully loaded vary depending on the browser version and file content."

File: docs/source/user/file_formats.rst [hunk 7]
```diff
@@ -129,6 +127,24 @@ browser and select the “Editor” item in the “Open With” submenu:
...
   </div>
 
+JupyterLab's powerful grid viewer can load files up to the maximum string
+size in the browser. Our experience testing the grid viewer on various browsers suggests that maximum string size varies with browser.
+Below is a table that reflects the maximum size we've successfully tested the grid viewer by browser:
+
++---------+----------+
+| Browser | Max Size |
++=========+==========+
+| Firefox | ~200MB   |
```
Comment (jasongrout) [inline docs/source/user/file_formats.rst]: I'm running the tests again to make sure. I get 250MB for firefox. I would also delete the tilde, since I'm testing more exact sizes now.

File: docs/source/user/file_formats.rst [hunk 8]
```diff
@@ -127,23 +127,21 @@ browser and select the “Editor” item in the “Open With” submenu:
...
+Below is a table that shows the sizes of the largest test files we successfully opened in each browser we support:
 
 +---------+----------+
 | Browser | Max Size |
 +=========+==========+
-| Firefox | ~200MB   |
+| Firefox |  250MB   |
 +---------+----------+
-| Chrome  | ~700MB   |
+| Chrome  |  700MB   |
```
Comment (jasongrout) [inline docs/source/user/file_formats.rst]: More data: Chrome: 730MB

File: docs/source/user/file_formats.rst [hunk 9]
```diff
@@ -127,23 +127,21 @@ browser and select the “Editor” item in the “Open With” submenu:
...
 +---------+----------+
-| Safari  | ~1.5GB   |
-+---------+----------+
-| IE      | ~1.5GB   |
+| Safari  |  1.5GB   |
```
Comment (jasongrout) [inline docs/source/user/file_formats.rst]: 1.7GB

File: docs/source/user/file_formats.rst [hunk 10]
```diff
@@ -127,23 +127,21 @@ browser and select the “Editor” item in the “Open With” submenu:
...
 +---------+----------+
 
-These numbers are intented to be conservative estimates.  Additionally, string encoding will also affect the ability to load CSV-like files on JupyterLab.
+These numbers are conservative estimates.
```
Comment (jasongrout) [inline docs/source/user/file_formats.rst]: They're not conservative anymore :). I tested right up until the browser tab crashed or I got an error.

File: docs/source/user/file_formats.rst (hunk 6 above)
Comment (jzf2101) [reply docs/source/user/file_formats.rst]: Kept first sentence and changed second to: "The actual maximum size of files that can be successfully loaded will vary depending on the browser version and file content."

**Comment by jasongrout** [comment]
It also appears we don't recognize the .tsv suffix: https://github.com/jupyterlab/jupyterlab/blob/master/packages/csvviewer-extension/src/index.ts

**Comment by jzf2101** [comment]
Should make that an issue? That makes no sense if we accept tab-separated files?

**Comment by jasongrout** [comment]
I agree it should be fixed. I'm just describing what the code does so our docs here aren't wrong.

File: docs/source/user/file_formats.rst [hunk 11]
```diff
@@ -101,25 +101,26 @@ submenu:
 
 .. _csv:
 
-CSV
-~~~
+Delimiter-separated Values
+~~~~~~~~~~~~~~~~~~~~~~~~~~
 
--  File extension: ``.csv``
+-  File extension: ``.csv``,
```
Comment (jasongrout) [inline docs/source/user/file_formats.rst]: Delete trailing comma

**Comment by jasongrout** [comment]
One final note about a trailing comma, and I think we are good!

**Comment by jasongrout** [comment]
Thank you very much! I think this is much better than what we had!

**Comment by AbdealiJK** [comment]
Quick question on this ... Wouldn't the maximum size of file be more dependent on the computer (RAM and other limitations) rather than the Browser ?

Other observations:
 - Opening ~800MB CSV in Chrome fails with a "Aw snap" (Note, after opening this, I cannot open the chrome tab again as every time when I open the tab it keeps giving a Aw Snap :D) - Opened https://github.com/jupyterlab/jupyterlab/issues/4239
 - Opening ~2.5GB CSV in Chrome gives a 500 internal error

**Comment by lejabque** [comment]
Same problem. When I try to open my notebook, RAM usage of browser is going to 4gb. 
There is no problem with that notebook in good old Jupyter

**Comment by jasongrout** [comment]
> Same problem. When I try to open my notebook, RAM usage of browser is going to 4gb. And then it giving an Aw Snap
> There is no problem with that notebook in good old Jupyter

This issue is about fixing opening large CSV files, not notebooks.

How big is your notebook? Can you file a new issue about opening large notebooks? Also, what browser, version, and OS are you using?

**Comment by lejabque** [comment]
Opened new issue here
#4269

**Comment by jasongrout** [comment]
Thanks!
